src/manifold/src/smoothing.cpp

```diff
@@ -308,6 +191,296 @@ struct InterpTri {
     pos = HNormalize(posH);
   }
 };
+
+class Partition {
+ public:
+  glm::ivec3 idx;
+  glm::ivec3 sortedDivisions;
+  Vec<glm::vec3> vertBary;
+  Vec<glm::ivec3> triVert;
+
+  int InteriorOffset() const {
+    return sortedDivisions[0] + sortedDivisions[1] + sortedDivisions[2];
+  }
+
+  int NumInterior() const { return vertBary.size() - InteriorOffset(); }
+
+  static Partition GetPartition(glm::ivec3 divisions) {
+    glm::ivec3 sortedDiv = divisions;
+    glm::ivec3 triIdx = {0, 1, 2};
+    if (sortedDiv[2] > sortedDiv[1]) {
+      std::swap(sortedDiv[2], sortedDiv[1]);
+      std::swap(triIdx[2], triIdx[1]);
+    }
+    if (sortedDiv[1] > sortedDiv[0]) {
+      std::swap(sortedDiv[1], sortedDiv[0]);
+      std::swap(triIdx[1], triIdx[0]);
+      if (sortedDiv[2] > sortedDiv[1]) {
+        std::swap(sortedDiv[2], sortedDiv[1]);
+        std::swap(triIdx[2], triIdx[1]);
+      }
+    }
+
+    Partition partition = GetCachedPartition(sortedDiv);
+    partition.idx = triIdx;
+
+    return partition;
+  }
+
+  Vec<glm::ivec3> Reindex(glm::ivec3 tri, glm::ivec3 edgeOffsets,
+                          glm::bvec3 edgeFwd, int interiorOffset) const {
+    Vec<int> newVerts;
+    newVerts.reserve(vertBary.size());
+    glm::ivec3 triIdx = idx;
+    std::vector<int> outTri = {0, 1, 2};
+    if (idx[1] != Next3(idx[0])) {
+      triIdx = {idx[2], idx[0], idx[1]};
+      edgeFwd = glm::not_(edgeFwd);
+      std::swap(outTri[0], outTri[1]);
+    }
+    for (const int i : {0, 1, 2}) {
+      newVerts.push_back(tri[triIdx[i]]);
+    }
+    for (const int i : {0, 1, 2}) {
+      const int n = sortedDivisions[i] - 1;
+      int offset = edgeOffsets[idx[i]] + (edgeFwd[idx[i]] ? 0 : n - 1);
+      for (int j = 0; j < n; ++j) {
+        newVerts.push_back(offset);
+        offset += edgeFwd[idx[i]] ? 1 : -1;
+      }
+    }
+    const int offset = interiorOffset - newVerts.size();
+    for (int i = newVerts.size(); i < vertBary.size(); ++i) {
+      newVerts.push_back(i + offset);
+    }
+
+    Vec<glm::ivec3> newTriVert(triVert.size());
+    for (int i = 0; i < triVert.size(); ++i) {
+      for (const int j : {0, 1, 2}) {
+        newTriVert[i][outTri[j]] = newVerts[triVert[i][j]];
+      }
+    }
+    return newTriVert;
+  }
+
+ private:
+  static std::unordered_map<glm::ivec3, Partition> cache;
```

Comment:
I was expecting I would need a bit of help making this safe in parallel, but I'm surprised I'm having trouble in single-threaded mode.

Reply:
interesting, will have a look